stl/inc/xutility

```diff
@@ -5284,9 +5284,75 @@ _NODISCARD constexpr _Iter_diff_t<_InIt> _Count_pr(_InIt _First, const _InIt _La
     return _Count;
 }
 
+enum class _TrimResult : unsigned char { _Continue, _Break, _ReturnFalse, _ReturnTrue };
+
+template <class _BidIt1, class _BidIt2, class _Pr>
+_NODISCARD _CONSTEXPR20 _TrimResult _Trim_equal(
+    _BidIt1& _First1, _BidIt1& _Back1, _BidIt2& _First2, _BidIt2& _Back2, _Pr _Pred) {
+    _STL_INTERNAL_CHECK(_First1 != _Back1);
+    _STL_INTERNAL_CHECK(_STD distance(_First1, _Back1) == _STD distance(_First2, _Back2));
+    if (_Pred(*_First1, *_First2)) {
+        do {
+            ++_First1;
+            ++_First2;
+            if (_First1 == _Back1) {
+                // only one element is left
+                return _Pred(*_First1, *_First2) ? _TrimResult::_ReturnTrue : _TrimResult::_ReturnFalse;
+            }
+        } while (_Pred(*_First1, *_First2));
+    } else {
+        if (!_Pred(*_Back1, *_Back2)) {
+            // nothing to trim, break
+            return _TrimResult::_Break;
+        }
+        --_Back1;
+        --_Back2;
+    }
+
+    while (_Pred(*_Back1, *_Back2)) {

```

Comment:
Although this looks correct to me, I believe that it can call the predicate redundantly. This happens when the first pair of elements are non-equal (i.e. the `if (_Pred(*_First1, *_First2))` on line 5294 returns `false`), the last pair of elements are equal (i.e. `if (!_Pred(*_Back1, *_Back2))` on line 5304 returns `false`), and we trim the last pair of elements. If we started with only two elements in each sequence, then `_First1 == _Back1` here at line 5312, so this `while (_Pred(*_Back1, *_Back2))` ends up testing the first pair of elements (that we tested on line 5294).

It seems to me that the partial duplication of trimming logic on lines 5304-5309 (a one-time trimming operation) and the `while` loop on line 5312-5315 leads to this confusion. Would it be possible to replace this with a single `while` loop, trimming elements at the back, that counts how many elements were trimmed, and returns `_TrimResult::_Break` when `0` elements are trimmed? That would uniformly handle the elements, and I believe it could avoid any redundant applications of the `_Pred`.